# Introduction
---

**pytorch**是python的一個程式庫，就像是處理天文數據會用astropy那樣，**pytorch**是專門拿來開發AI的工具，可想像成可以被GPU加速的numpy

有二大特點，使其能夠很好的作為深度學習的框架：
1. 利用 **張量(tensor)** 作為處理資料的基本單位，並且可被丟到GPU上面做數學運算
2. **自動微分引擎(Autograd)**：我們在寫程式只要定義好forward要怎麼算，系統就會自動幫我們紀錄路徑，再透過連鎖律算出所有參數的微分，從而修正模型參數

# 前置準備
---

函式庫更新可能追不上python版本更新，所以建議使用比最新版本低1-2版本的python(我是用3.10)；有CUDA相容的GPU的建議就安裝支援GPU的版本，嘗試看看用GPU跑程式  
PS: CUDA是個運算平台，負責*指揮*GPU去看懂程式

關於要在terminal打甚麼指令安裝pytorch，因為每個人系統不一樣，因此請見官網：
`https://pytorch.org/`  
然後這邊採用PyTorch 2.4.0，所以建議把指令的torch改成`torch==2.4.0`

In [1]:
import torch

# 張量簡介
---
- 0階張量: scalar
- 1階張量: vector
- 2階張量: Matrix
- etc.

In [2]:
# create a tensor
tensor_0d = torch.tensor(1)
tensor_1d = torch.tensor([1, 2, 3])
tensor_2d = torch.tensor([[1, 2, 3],
                          [4, 5, 6]])

for i in range(3):
    if i == 0:
        tensor = tensor_0d
    elif i == 1:
        tensor = tensor_1d
    else:
        tensor = tensor_2d
    print(f"{i}D Tensor:", tensor,tensor.shape,"\n",sep="\n")  

0D Tensor:
tensor(1)
torch.Size([])


1D Tensor:
tensor([1, 2, 3])
torch.Size([3])


2D Tensor:
tensor([[1, 2, 3],
        [4, 5, 6]])
torch.Size([2, 3])




In [3]:
# from the output, we can see that the 2D tensor has shape (2, 3),
# if we want to reshape it to (3, 2), we can do as follows:

# reshape tensors(take 2D tensor as example)
tensor = tensor_2d

# there are two ways to reshape a tensor
# 1. using reshape() method
re_tensor_1 = tensor.reshape(3, 2)

# 2. using view() method
re_tensor_2 = tensor.view(3, 2)

print(re_tensor_1,"\n\n",re_tensor_2)

tensor([[1, 2],
        [3, 4],
        [5, 6]]) 

 tensor([[1, 2],
        [3, 4],
        [5, 6]])


In [4]:
# to transpose a tensor, using T attribute
tensor_1 = torch.tensor([[1, 2, 3],
                       [4, 5, 6],
                       [7, 8, 9]])
tensor_transposed_1 = tensor_1.T

print(tensor_1,"\n\n",tensor_transposed_1)

print("=======================")

# try anothor tensor
tensor_2 = torch.tensor([[1, 2, 3],
                         [4, 5, 6]])
tensor_transposed_2 = tensor_2.T

print(tensor_2, "\n\n", tensor_transposed_2)



tensor([[1, 2, 3],
        [4, 5, 6],
        [7, 8, 9]]) 

 tensor([[1, 4, 7],
        [2, 5, 8],
        [3, 6, 9]])
tensor([[1, 2, 3],
        [4, 5, 6]]) 

 tensor([[1, 4],
        [2, 5],
        [3, 6]])


In [5]:
# if we want to muptliply two tensors, there are two methods
tensor_a = torch.tensor([[1, 2, 3],
                         [4, 5, 6]])
tensor_b = torch.tensor([[1, 2],
                         [3, 4],
                         [5, 6]])

# 1. using .matmul()
tensor_c = tensor_a.matmul(tensor_b)
print(tensor_c)

print("=======================")

# 2. using @
tensor_c = tensor_a@tensor_b
print(tensor_c)

tensor([[22, 28],
        [49, 64]])
tensor([[22, 28],
        [49, 64]])


## 大家都要學會的招數

```.to()``` 可以讓張量運算指定到GPU上運行

In [6]:
x = torch.tensor([1, 2, 3])
x = x.to("cuda")            # Move tensor to GPU

In [68]:
tensor_1 = torch.tensor([1, 2, 3])
tensor_2 = torch.tensor([4, 5, 6])

# 利用.to()將張量移到GPU上面做運算看看
tensor_1 = tensor_1.to("cuda:0")
tensor_2 = tensor_2.to("cuda:0")
print(tensor_1 + tensor_2)

tensor([5, 7, 9], device='cuda:0')


### 大家跑比賽

In [7]:
import time
# 謝謝G學長教我怎麼算時間
N = 8000

tensor_a = torch.randn(N, N)
tensor_b = torch.randn(N, N)
# move to GPU
tensor_a_GPU = tensor_a.to("cuda")
tensor_b_GPU = tensor_b.to("cuda")

# 預熱 (Warm-up)
_ = tensor_a @ tensor_b 
_ = tensor_a_GPU @ tensor_b_GPU
torch.cuda.synchronize()        # 等待預熱完成

# ============================
# CPU 系列
# ============================
start = time.time()

result_CPU = tensor_a@tensor_b

end = time.time()

cpu_time_ms = (end - start) * 1000  # s -> ms

print(f"CPU time: {cpu_time_ms:.2f} ms")

# ============================
# GPU 系列
# ============================
# create events
start_GPU = torch.cuda.Event(enable_timing=True)
end_GPU = torch.cuda.Event(enable_timing=True)

start_GPU.record()

result_GPU = tensor_a_GPU@tensor_b_GPU

end_GPU.record()

# 等待GPU跑完
torch.cuda.synchronize()

gpu_time_ms = start_GPU.elapsed_time(end_GPU)

print(f"GPU time: {gpu_time_ms:.2f} ms")

CPU time: 5764.29 ms
GPU time: 850.86 ms


# Computation Graph & Autograd(計算圖和自動微分)
---

- Computation Graph:  
當程式碼在向前傳遞(Forward)時，pytorch會將計算過程紀錄下來，以便之後調整參數，簡單來說，模型不會忘記之前做過的事情，方便之後回溯過程調整參數
- Autograd:  
調整參數的軍師，透過連鎖率計算出參數`調整的方向和大小`，再透過Optimizer(武官)去做調整

兩者的關聯:  
Computation Graph提供了運算的路徑，讓Autograd知道怎麼往回走，一步步算出每個節點對最終錯誤的影響(微分)，最後把這些建議交給Optimizer來更新參數

### 手搓神經網路

先自己定義好欲預測答案(y)、輸入特徵(x)、權重(w)、神經元的bias(b)；如果現在在上實驗物理學二的話，y就是實驗數據點、x是實驗條件。  
神經網路的連接都是透過數學方程式連接的，利用線性方程式z=wx+b算出理論結果，並和實驗結果y比較，算出其微分作為誤差調整方向  
*請想到我們最愛的七周chi square fitting的訓練，現在透過autograd去自動計算梯度，讓 loss(類似reduced chi-square的角色)自動收斂到最小值*

In [8]:
import torch.nn.functional as F

y = torch.tensor([1.0])         # answer
x1 = torch.tensor([1.1])        # 輸入特徵
w1 = torch.tensor([2.2])        # weights
b = torch.tensor([0.0])         # bias unit

z = x1 * w1 + b         # input
a = torch.sigmoid(z)    # activation function and output

loss = F.binary_cross_entropy(a, y) # 計算損失
print(loss)

tensor(0.0852)


解釋一下上面的程式碼:  
- z是線性輸出的`原始分數`  
    - 看上面的例子: $z=w_{1} \cdot x_{1}+b$ $\rightarrow$ $1.1 * 2.2 + 0 = 2.42$
- a是激活輸出，這個是`機率值`
    - Sigmoid ~~乙狀結腸??~~ 是一個*激活函數*，公式為: $\sigma(z)=\frac{1}{1+e^{-z}}$
    - 一樣focus在上面的例子: $a = Sigmoid(z)$，將$z = 2.42$帶入後，$a = \frac{1}{1+e^{-2.42}}\approx 0.9183$
    - 經過Sigmoid函數後原始分數會被壓縮到0-1之間，進而得到`答案為y時的機率值`

現在來手算 loss驗證和print出來的答案是否相同:  
Binary Cross Entropy的公式: $Loss = -[y \cdot \ln(a) + (1-y) \cdot \ln(1-a)]$  
$\Rightarrow Loss = -1 \cdot \ln(a)$  
$\Rightarrow Loss = -\ln (0.9183) \approx -(-0.0852) = 0.0852$  
和print出來的一致

對於激活函數，如果想更了解可以參考這篇文章:[激活函數](https://ithelp.ithome.com.tw/m/articles/10276865)

In [9]:
### 加入自動微分
import torch.nn.functional as F
from torch.autograd import grad

y = torch.tensor([1.0])
x1 = torch.tensor([1.1])
w1 = torch.tensor([2.2], requires_grad=True)    # requires_grad=True目的在告訴pytorch這個變數是參數(權重)，不是其他nobody
b = torch.tensor([0.0], requires_grad=True)

z = x1 * w1 + b 
a = torch.sigmoid(z)

loss = F.binary_cross_entropy(a, y)

# 自動微分(法一): grad手動版本
# ============================================
grad_L_w1 = grad(loss, w1, retain_graph=True)   # 求loss和w1的微分
grad_L_b = grad(loss, b, retain_graph=True)
# ============================================
# 沒有儲存到 w1.grad 這個屬性裡面

print(grad_L_w1)
print(grad_L_b)

(tensor([-0.0898]),)
(tensor([-0.0817]),)


這段程式碼只多了`grad`這個部分，也是自動微分，我們還是可以手算看看去驗證與print的結果是否相同  
- 已知路徑: $w,b \rightarrow z \rightarrow a \rightarrow Loss$
- 已知參數: 
    - $x = 1.1$
    - $y = 1.0$
    - $a = 0.9183$
- 施魔法找到$\frac{\partial Loss}{\partial z} = a-y$
    - $\frac{\partial Loss}{\partial a} = \frac{a-y}{a(1-a)}$
    - $\frac{\partial a}{\partial z} = a(1-a)$
    - $\frac{\partial Loss}{\partial z} = \frac{\partial Loss}{\partial a} \cdot \frac{\partial a}{\partial z} = \frac{a-y}{a(1-a)} \cdot a(1-a) = a-y$

找答案: 
1. $\frac{\partial Loss}{\partial w_1}=(a-y) \cdot x$  
$\Rightarrow (0.9183-1.0) \cdot 1.1 = -0.08987$
2. $\frac{\partial Loss}{\partial b}=(a-y) \cdot 1$
$\Rightarrow 0.9183-1.0 = -0.0817$

In [10]:
# 自動微分(法二): loss.backward()全自動廣播
# 透過反向傳播，自動把計算圖中所有需要微分的參數都計算過，並填入各自.grad的屬性
loss.backward()

# 確認數值和手動版一致
print(w1.grad)
print(b.grad)

tensor([-0.0898])
tensor([-0.0817])


### 模組版本

上面都是原理方面的程式碼，真正要實戰的話會用下面這種模組版本的寫法

In [11]:
class MyLLM(torch.nn.Module):

    # 定義零件(層)
    def __init__(self):
        super().__init__()                      # 進行初始化神經網路的部分
        self.layer1 = torch.nn.Linear(10, 5)    # 定義第一層: 輸入10個特徵輸出5個特徵

    # 定義資料流向
    def forward(self, x):
        return self.layer1(x)                   # 資料向前傳遞；輸入x值後執行第一層

- torch.nn.Linear(10, 5) $\rightarrow$ 自動生成10*5的矩陣(w)和長度為5的向量(b)
- self.layer1 $\rightarrow$ 自動執行矩陣乘法 + bias  
這只是定義模型的架構，還沒**執行**，所以沒加入 loss.backward()算微分

# 建構神經網路
---
我們現在從「單顆小神經元」進入到cp大亂鬥的多層神經網路了

In [12]:
# 有兩個隱藏層的多層感知器
class NeuralNetwork(torch.nn.Module):
    def __init__(self, num_inputs, num_outputs): 
        super().__init__()

        # 用Sequential把所以layer打包
        self.layers = torch.nn.Sequential(

            # 1st hidden layer
            torch.nn.Linear(num_inputs, 30),
            torch.nn.ReLU(),                    # add activation function

            # 2nd hidden layer
            torch.nn.Linear(30, 20),            # 頭尾數字一樣
            torch.nn.ReLU(),

            # output layer
            torch.nn.Linear(20, num_outputs),
        )
     # 有人說貨出的去，人進得來，高雄發大財
     # 所以資料進去，答案出來，模型不會壞
    def forward(self, x):
        logits = self.layers(x)
        return logits

1. 先前在`MyLLM`是手動把零件一個一個拿出來(layer1(x) $\rightarrow$ layer2(x))；這邊透過用`nn.Sequential`這個物件，進行forward時，資料就會自動從第一層一路過關斬將跑到最後一層
2. ReLU是一種**激活函數**:
    - if $x>0$, 輸出$x$
    - if $x \le 0$, 輸出$0$  
如果只有Linear，不管我疊了幾千幾百萬層，那還是一條直線；透過加入ReLU，可以更靈活地去fitting數據(加入non-linear)
3. 還滿像跑大隊接力的，記得從上一層接到多少輸出下一層輸入就要有多少

In [13]:
# 建立一個有三個input，2個output的模型
### 大注意，這裡的input是指一筆資料被提取3個特徵，而不是三筆資料
# if u want to fix the initial weights, u can set a random seed
'''
torch.manual_seed(123)
'''
model = NeuralNetwork(3, 2)

In [14]:
# now focus on the number of its parameters
num_params = sum(p.numel() for p in model.parameters() if p.requires_grad)
print("Total number of trainable model parameters:", num_params)
print("==========================")
print(model.layers[0].weight)       # cuz we just want to check the first layer, so it'll print out the 3*30 matrix 
print("==========================")
print(model.layers[0].weight.shape) # check its shape

Total number of trainable model parameters: 782
Parameter containing:
tensor([[ 0.4995,  0.3526,  0.2445],
        [ 0.0528,  0.0050,  0.0431],
        [ 0.3842, -0.3361,  0.2907],
        [ 0.1429,  0.3798, -0.1017],
        [-0.2314,  0.2047, -0.0837],
        [-0.0235, -0.1529,  0.2704],
        [-0.1523, -0.3388, -0.1225],
        [ 0.4549, -0.1404, -0.3527],
        [ 0.3878, -0.0515,  0.2490],
        [ 0.5303,  0.4374, -0.5093],
        [ 0.5085,  0.5605,  0.0182],
        [ 0.3415,  0.5624, -0.2308],
        [-0.3189,  0.0082,  0.5325],
        [ 0.4077,  0.2762,  0.4231],
        [-0.3524,  0.1576,  0.5655],
        [-0.1585,  0.0330, -0.1287],
        [-0.4158,  0.4404, -0.2835],
        [ 0.5209,  0.3664, -0.1358],
        [ 0.0308, -0.0964,  0.0096],
        [ 0.4722, -0.0507,  0.2848],
        [-0.5328, -0.5637,  0.2192],
        [-0.1886,  0.2999,  0.4202],
        [-0.3728,  0.5170, -0.0233],
        [-0.1463,  0.0265, -0.0029],
        [ 0.3898, -0.4073,  0.0014],
     

這個程式碼做三件事: 
1. 印出參數數量  
    同樣的用手算的方式驗證看看，大家腦中應該有神經網路的圖(詳情見紅魚0114的slide，或是上網找應該隨便都找得到)  
    (1) 第一層: 3個input、30個output:  
    - 權重: $3\cdot 30=90$
    - bias: $30$
    (2) 第二層: 30個input、20個output: 
    - 權重: $30\cdot 20=600$
    - bias: $20$
    (3) 第三層: 20個input、2個output: 
    - 權重: $20\cdot 2 = 40$
    - bias: 2  
    加總: $90+30+600+20+40+2=782 \rightarrow$和print相同
2. 印出第一層的權重矩陣  
    因為現在還沒做訓練，所以在跑`NeuralNetwork`這個模組後他會先隨機給一些小數字方便日後訓練(重跑cell47、48就可以驗證這件事)；而第一層有3個input、30個output，所以會是3*30的矩陣
3. 欸不是，不是3\*30嗎?為什麼是30\*3的矩陣呢?  
    因為在pytorch的nn.Linear中，權重的儲存格式是`[out_features, in_features]`，代表說pytorch在運算時會自動幫我們轉置；**因為運算公式是$y_{1}=x \cdot W^{T}_{1}$，本身就包含矩陣的轉置，而我們要和輸入相乘，要符合矩陣乘法規則**  
    e.g. 一筆資料被提取3個特徵(這裡的input) $\rightarrow [1, 3]$，能和$[30, 3]^{T}=[3, 30]$相乘

In [15]:
X = torch.rand((1, 3))     # input data(random)
out = model(X)
print(out)

# 不進行訓練或反向傳播，可節省記憶體
with torch.no_grad():
    out1 = model(X)
    out2 = torch.softmax(model(X), dim=1)    # 用softmax標準化，輸出即為機率
print(out1)              # output為原始數值，未經標準化
print(out2)

tensor([[-0.0225, -0.1325]], grad_fn=<AddmmBackward0>)
tensor([[-0.0225, -0.1325]])
tensor([[0.5275, 0.4725]])


透過`.rand`隨機選擇三個input data(一筆`特徵向量為3維度`的資料)，把這個資料餵給`model`後存在`out`裡面並印出來  
out是計算圖的中間儲存點，還在等著完成後面的反向傳播  
out1單純是印出原始分數(Logits)(for 確認這邊結果一樣)，加入`torch.no_grad()`切斷與計算圖的連結  
out2是透過用`softmax`激活函數將原始分數壓成機率

# Dataset & DataLoader
---

Dataset: 切資料的模組，告訴模型資料在哪裡、共有多少資料  
Dataloader: 端資料給模型吃

In [16]:
# 自己建立訓練集和測試集

# =====================================
# 訓練集
    # 5個訓練樣本，每個樣本包含2個特徵
X_train = torch.tensor([                
    [-1.2, 3.1],
    [-0.9, 2.9],
    [-0.5, 2.6],
    [2.3, -1.1],
    [2.7, -1.5]
    ])
    # 5個訓練樣本對應的標籤
y_train = torch.tensor([0, 0, 0, 1, 1])
# =====================================

# =====================================
# 測試集
X_test = torch.tensor([
    [-0.8, 2.8],
    [2.6, -1.6],
])
y_test = torch.tensor([0, 1])
# =====================================

In [17]:
from torch.utils.data import Dataset

class ToyDataset(Dataset):
    def __init__(self, X, y):           # 初始設定，將準備好的資料(X)(特徵)和答案(y)輸入
        self.features = X
        self.labels = y

    def __getitem__(self, index):       # 取得某一筆資料，告訴index幾，取出幾號資料
        one_x = self.features[index] 
        one_y = self.labels[index]
        return one_x, one_y
    
    def __len__(self):                  # 資料總長度
        return self.labels.shape[0]
    
train_ds = ToyDataset(X_train, y_train) # 訓練集資料的切割
test_ds = ToyDataset(X_test, y_test)    # 測試集資料的切割

# 印出訓練集的資料切割，確認切成5份
print(len(train_ds))

5


**Dataset(資料集):**  
在這個部分，我們將把資料準備好
1. `__init__`:  
    把訓練集的特徵向量(X)和其帶的標籤(答案)(y)儲存在這個類別中
2. `__getitem__`:  
    取出特定的、第i份的資料(為了後面dataloader洗牌)
3. `__len__`:  
    總共有幾份資料
    
也就是一大塊豆干(所有資料集合)切成一盤一盤的豆干絲(內容包含一筆資料(特徵向量)一筆答案)  
~~為什麼是豆干，因為原本是切肉但覺得不好~~  
可以去[Data_Preprocessing_and_Embeddings.ipynb](https://github.com/aqingo3o/1004-LLM/blob/08493d94657f2f378bf6cdcc44331eabfe526747/tutorial/RF/Data_Preprocessing_and_Embeddings.ipynb)看對於自然語言文本的dataset

所以對於訓練集來說，總共有五筆:  
| 特徵向量 (X) | 對應標籤 (y) |
| :---: | :---: |
| `[-1.2, 3.1]` | 0 |
| `[-0.9, 2.9]` | 0 |
| `[-0.5, 2.6]` | 0 |
| `[2.3, -1.1]` | 1 |
| `[2.7, -1.5]` | 1 |

dataset會把他切成五筆訓練資料對，每行一對(包含一個特徵向量一個標籤)

In [61]:
from torch.utils.data import DataLoader

#個別把訓練集和測試集切好的資料餵給模型

train_loader = DataLoader(
    dataset=train_ds,       # dataset選擇訓練集切完的資料
    batch_size=2,           # 每次端幾筆資料給model吃
    shuffle=True,           # 洗牌，每次取的資料順序不同，防止模型背答案
)

test_loader = DataLoader(
    dataset=test_ds,
    batch_size=2,
    shuffle=False,          # 對於測試集只是要打分數沒必要洗牌
     num_workers=0 
)

DataLoader的參數:  
1. dataset:  
    選擇dataset切好的資料(訓練集or測試集)
2. batch_size:  
    每次餵幾筆資料給模型，以訓練集當例子來說，第一次迴圈他會吃進去五筆訓練資料對中的兩筆訓練資料對
3. shuffle:  
    防止模型被資料對的順序影響或是背答案，所以在訓練時打亂訓練資料對的順序；而測試集只是要看有沒有訓練好(能否回答正確)，因此不用洗牌
4. drop_last(這邊是False):  
    總共五筆資料，每口吃兩筆資料的話，會有一筆資料烙單，如果設定drop_last=True時，會直接將烙單的資料丟掉

討論`num_workers=0`的情況  
此參數被設定為0時，資料載入會在主處理程序中進行，而不是在單獨的工作處理程序中進行  
若設為1或更大的數字，可以平行處理事情，更有效的利用系統資源


In [62]:
# 把每個Batch印出來更直觀的感受model一口吃了甚麼東西
for idx, (x, y) in enumerate(train_loader):
    print(f"Batch {idx+1}:", f"feature: {x}", f"label: {y}", sep="\n")
    print("================")

# 重複執行會發現每次吃一口裡面的資料不一定相同(洗牌的作用)

Batch 1:
feature: tensor([[-0.9000,  2.9000],
        [-1.2000,  3.1000]])
label: tensor([0, 0])
Batch 2:
feature: tensor([[ 2.7000, -1.5000],
        [ 2.3000, -1.1000]])
label: tensor([1, 1])
Batch 3:
feature: tensor([[-0.5000,  2.6000]])
label: tensor([0])


In [63]:
# 注意到最後一個batch只有1筆資料，如果訓練週期中最後一個批次遠小於其他批次可能會影響訓練的收斂性
# 設定drop_last=True可以捨棄最後一個batch

train_loader = DataLoader(
    dataset=train_ds,
    batch_size=2,          
    shuffle=True,          
    drop_last=True          #捨棄最後一個Batch
)

# 發現最後一個被丟掉了
for idx, (x, y) in enumerate(train_loader):
    print(f"Batch {idx+1}:", x, y, sep="\n")
    print("================")

Batch 1:
tensor([[ 2.7000, -1.5000],
        [ 2.3000, -1.1000]])
tensor([1, 1])
Batch 2:
tensor([[-0.5000,  2.6000],
        [-1.2000,  3.1000]])
tensor([0, 0])


# 實戰一個 訓練迴圈
---

In [66]:
import torch.nn.functional as F

torch.manual_seed(123)  # 設定種子碼固定一下，不然每次跑都不一樣
model = NeuralNetwork(num_inputs=2, num_outputs=2)      # 設定模型是上面設定過兩層的神經網路，input是2維特徵向量、output是2種標籤
optimizer = torch.optim.SGD(model.parameters(), lr=0.5) # 優化器?這是啥

num_epochs = 3      # 代表要訓練多少個週期(要把所有訓練資料跑三次)
                    # Epoch: 總共要訓練幾次，這邊設定三次，代表說巨大豆干(所有資料)吃完後會再刷新一次，直到吃了三次後結束

for epoch in range(num_epochs):
    model.train()   # 訓練模式on
    for batch_idx, (features, labels) in enumerate(train_loader):
        # step 1: 猜答案(Forward pass)
        logits = model(features)                # 資料餵進去，算出logits(原始分數)

        # step 2: 算分數(Calculate loss)
        loss = F.cross_entropy(logits, labels)  # 比較猜出來的原始分數和標籤差多少

        # step 3: 將前一輪的梯度設為0(Zero gradients)
        optimizer.zero_grad()                   # 不燃pytorch會把兩次梯度相加...(貼心)
        # step 4: 尋找修正方向(Backward pass)(自動微分)
        loss.backward()                         # 逆向傳播：算出每個參數該怎麼調，誤差才會變小
        # step 5: 實際修正(Optimizer step)
        optimizer.step()                        # 依照step 4算出的方向，實際更新模型參數 (w = w - lr * gradient)
        
        print(f"Epoch: {epoch+1:03d}/{num_epochs:03d}"
              f" | Batch {batch_idx:03d}/{len(train_loader):03d}"
              f" | Train/Val Loss: {loss:.2f}")
        
    model.eval()    # 評估模式on

Epoch: 001/003 | Batch 000/002 | Train/Val Loss: 0.75
Epoch: 001/003 | Batch 001/002 | Train/Val Loss: 0.65
Epoch: 002/003 | Batch 000/002 | Train/Val Loss: 0.44
Epoch: 002/003 | Batch 001/002 | Train/Val Loss: 0.13
Epoch: 003/003 | Batch 000/002 | Train/Val Loss: 0.03
Epoch: 003/003 | Batch 001/002 | Train/Val Loss: 0.00


銘記這五大步驟: 
1. 算結果: 將資料丟進模型，算出預測值
2. 算分數: 比較預測值和標準答案差多少
3. 清空紀錄: 把上一次的修正建議忘記，避免干擾這一次的
4. 找方向: 叫助手回溯過程，算出這一次要怎麼修正
5. 更新: 依照修正建議，實際去調整模型的參數

之後在做訓練時都是尊循這五個步驟在進行

In [70]:
model.eval()
with torch.no_grad():           # 關閉記錄功能(因為現在只要預測)
    outputs = model(X_train)    # 寫考古題
print(outputs)

tensor([[ 2.8569, -4.1618],
        [ 2.5382, -3.7548],
        [ 2.0944, -3.1820],
        [-1.4814,  1.4816],
        [-1.7176,  1.7342]])


印出考古題的原始分數，所以模型可能就會知道，喔齁齁如果標籤是`0`時，第一維度會比第二維度還大；標籤是`1`時，第二維度會比第一維度還大；重點在於**比大小**

In [ ]:
torch.set_printoptions(sci_mode=False)      # 關掉科學記號
probas = torch.softmax(outputs, dim=1)      # 透過softmax這個激活函數，把分數轉機率
print(probas)

predictions = torch.argmax(probas, dim=1)   # 回傳機率最大的類別
print(predictions)

tensor([[    0.9991,     0.0009],
        [    0.9982,     0.0018],
        [    0.9949,     0.0051],
        [    0.0491,     0.9509],
        [    0.0307,     0.9693]])
tensor([0, 0, 0, 1, 1])


`dim=1`: 資料橫向掃描，找到每一列最大值是哪個位置  
`dim=0`: 直的看(跨資料比較，沒意義)

In [ ]:
# 可以直接把argmax直接運用在logits上，就不用softmax這步了
predictions = torch.argmax(outputs, dim=1)
print(predictions)

tensor([0, 0, 0, 1, 1])


In [25]:
# 確認是否和答案一樣
predictions == y_train

tensor([True, True, True, True, True])

In [26]:
#利用torch.sum計算正確的數量
torch.sum(predictions == y_train)

tensor(5)

In [ ]:
# 把上述預測準確率的過程更通用，所以把他打包成compute_accuracy函式
def compute_accuracy(model, dataloader):
    model = model.eval()
    correct = 0.0       # 目前答對幾題
    total_examples = 0  # 目前總共考幾題
    
    # dataloader部分
    for idx, (features, labels) in enumerate(dataloader):
        
        with torch.no_grad():   # 預測答案
            logits = model(features)                # 猜答案(原始分數)
        predictions = torch.argmax(logits, dim=1)   # 轉成所預測標籤

        compare = labels == predictions         # 對答案進行比對、累加
        correct += torch.sum(compare)
        total_examples += len(compare)

    return (correct / total_examples).item()    # 回傳正確率(答對題數/總題數)


In [28]:
print(compute_accuracy(model, train_loader))
print(compute_accuracy(model, test_loader))

1.0
1.0


# After care
---

In [29]:
# 先進行儲存的動作，存到硬碟裡面
torch.save(model.state_dict(), "model.pth")

In [69]:
# 從硬碟讀取已儲存的模型參數
model = NeuralNetwork(2, 2)                     # 建立一個一樣架構的空模型給模型住進去，(2,2)需要和當初訓練的模型一致
model.load_state_dict(torch.load("model.pth"))  # 讀檔，就不用重新訓練了，可以直接預測
# 重新載入時預設是訓練模式(train())，記得如果只是要預測，要開成eval()

C:\Users\USER\AppData\Local\Temp\ipykernel_14712\1238848358.py:3: FutureWarning: You are using `torch.load` with `weights_only=False` (the current default value), which uses the default pickle module implicitly. It is possible to construct malicious pickle data which will execute arbitrary code during unpickling (See https://github.com/pytorch/pytorch/blob/main/SECURITY.md#untrusted-models for more details). In a future release, the default value for `weights_only` will be flipped to `True`. This limits the functions that could be executed during unpickling. Arbitrary objects will no longer be allowed to be loaded via this mode unless they are explicitly allowlisted by the user via `torch.serialization.add_safe_globals`. We recommend you start setting `weights_only=True` for any use case where you don't have full control of the loaded file. Please open an issue on GitHub for any issues related to this experimental feature.
  model.load_state_dict(torch.load("model.pth"))  # 讀檔，就不用重新訓練了

<All keys matched successfully>

In [31]:
# 確認環境可不可以使用GPU
print(torch.cuda.is_available())

True


In [34]:
# 將訓練迴圈用GPU跑

torch.manual_seed(123)
model = NeuralNetwork(num_inputs=2, num_outputs=2)
#---------------------------------------------------------#
device = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")
model = model.to(device)
#---------------------------------------------------------#
optimizer = torch.optim.SGD(model.parameters(), lr=0.5)

num_epochs = 3

for epoch in range(num_epochs):
    model.train()
    for batch_idx, (features, labels) in enumerate(train_loader):
#---------------------------------------------------------#
        features = features.to(device)
        labels = labels.to(device)
#---------------------------------------------------------#        
        logits = model(features)
        loss = F.cross_entropy(logits, labels)

        optimizer.zero_grad()
        loss.backward()
        optimizer.step()


        print(f"Epoch: {epoch+1:03d}/{num_epochs:03d}"
              f" | Batch {batch_idx:03d}/{len(train_loader):03d}"
              f" | Train/Val Loss: {loss:.2f}")
        
    model.eval()    # 評估模式on

Epoch: 001/003 | Batch 000/002 | Train/Val Loss: 0.75
Epoch: 001/003 | Batch 001/002 | Train/Val Loss: 0.65
Epoch: 002/003 | Batch 000/002 | Train/Val Loss: 0.44
Epoch: 002/003 | Batch 001/002 | Train/Val Loss: 0.13
Epoch: 003/003 | Batch 000/002 | Train/Val Loss: 0.03
Epoch: 003/003 | Batch 001/002 | Train/Val Loss: 0.00
